### PR TITLE
[simulation] enable the platform assert

### PR DESCRIPTION
**Pull request by zhanglongxia**
When the ot-rcp enters the assert state, the host crashes and exits. But the ot-rcp still runs in the dead loop and becomes an orphan process.

This commit enables the platform assert so that the ot-rcp can automatically exit when entering the assert state.

### CODE REVIEW

File: examples/platforms/simulation/misc.c [hunk 1]
```diff
@@ -113,3 +113,12 @@ otPlatMcuPowerState otPlatGetMcuPowerState(otInstance *aInstance)
...
 otError otPlatLogCrashDump(void) { return OT_ERROR_NONE; }
 #endif
+
+void otPlatAssertFail(const char *aFilename, int aLineNumber)
+{
+    otLogCritPlat("assert failed at %s:%d", aFilename, aLineNumber);
+
+    // For debug build, use assert to generate a core dump
+    assert(false);
+    exit(1);
```
Comment (Irving-cl) [inline examples/platforms/simulation/misc.c:123]: Is `exit` needed after `assert`?

**Review by Irving-cl — APPROVED** [review]
LGTM

File: tests/unit/test_platform.cpp [hunk 2]
```diff
@@ -967,4 +967,5 @@ OT_TOOL_WEAK void otPlatDnssdStopIp4AddressResolver(otInstance *aInstance, const
 OT_TOOL_WEAK otError otPlatLogCrashDump(void) { return OT_ERROR_NONE; }
 #endif
 
+void otPlatAssertFail(const char *, int) {}
```
Comment (abtink) [inline tests/unit/test_platform.cpp:970]: Let's declare it here as `OT_TOOL_WEAK` so specific tests can override it and provide their own if they want.

**Review by abtink — APPROVED** [review]
LGTM. Thanks.
One small suggestion below:

File: examples/platforms/simulation/misc.c (hunk 1 above)
Comment (zhanglongxia) [reply examples/platforms/simulation/misc.c:123]: The `assert()` does nothing if the ` NDEBUG` is defined. We need to call `exit()` in this case.

> If the macro NDEBUG is defined at the moment <assert.h> was last included, the macro assert() generates no code, and hence does nothing at all.

Refer to https://man7.org/linux/man-pages/man3/assert.3.html

**Review by jwhui — APPROVED** [review]

**Pull request merged**